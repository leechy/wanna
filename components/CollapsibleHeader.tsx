import {
  Text,
  StyleSheet,
  View,
  useColorScheme,
  TouchableOpacity,
} from 'react-native';
import Feather from '@expo/vector-icons/Feather';
import { Colors } from '@/constants/Colors';

interface CollapsibleHeaderProps {
  title: string;
  items?: number;
  color?: string;
  isOpen: boolean;
  onToggle: () => void;
}

export function CollapsibleHeader({
  title,
  items,
  color,
  isOpen,
  onToggle,
}: CollapsibleHeaderProps) {
  const theme = useColorScheme() ?? 'light';
  const titleColor =
    color ?? (theme === 'light' ? Colors.light.text : Colors.dark.text);

  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={styles.heading}
        onPress={onToggle}
        activeOpacity={0.4}
      >
        <Feather
          name="chevron-right"
          size={24}
          color={theme === 'light' ? Colors.light.text : Colors.dark.text}
          style={[
            styles.chevron,
            { transform: [{ rotate: isOpen ? '90deg' : '0deg' }] },
          ]}
        />
        <Text style={[styles.title, { color }]}>{title}</Text>
        <Text style={[styles.number, { color }]}>{items}</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    margin: 0,
    padding: 0,
    borderRightWidth: 1,
    borderRightColor: 'transparent',
  },
  chevron: {
    width: 24,
    height: 24,
  },
  heading: {
    paddingTop: 18,
    paddingBottom: 12,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    gap: 4,
  },
  title: {
    fontFamily: 'Nunito',
    fontSize: 18,
    fontWeight: 'bold',
    lineHeight: 24,
  },
  number: {
    marginLeft: 8,
    fontFamily: 'Nunito',
    fontSize: 16,
    fontWeight: 'bold',
    opacity: 0.5,
  },
});
