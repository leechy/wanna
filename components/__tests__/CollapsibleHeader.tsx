import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { CollapsibleHeader } from '../CollapsibleHeader';

jest.mock('../../assets/symbols/chevron-right.svg', () => 'ChevronRightIcon');
jest.mock('../../assets/symbols/chevron-down.svg', () => 'ChevronDownIcon');

jest.mock('@/hooks/useThemeColor', () => ({
  useThemeColor: () => '#0000FF',
}));

describe('CollapsibleHeader', () => {
  it('renders title correctly', () => {
    const { getByText } = render(<CollapsibleHeader title="Test Title" isOpen={false} onToggle={() => {}} />);
    expect(getByText('Test Title')).toBeTruthy();
  });

  it('renders with correct color', () => {
    const { getByText } = render(
      <CollapsibleHeader title="Test Title" color="#FF0000" isOpen={false} onToggle={() => {}} />
    );
    const title = getByText('Test Title');
    expect(title.props.style).toContainEqual({ color: '#FF0000' });
  });

  it('is clickable when clickable is true', () => {
    const onToggleMock = jest.fn();
    const { getByTestId } = render(
      <CollapsibleHeader
        title="Test Title"
        isOpen={false}
        clickable={true}
        onToggle={onToggleMock}
        testID="collapsible-header"
      />
    );
    const header = getByTestId('collapsible-header');
    fireEvent.press(header);
    expect(onToggleMock).toHaveBeenCalled();
  });

  it('is not clickable when clickable is false', () => {
    const onToggleMock = jest.fn();
    const { getByTestId } = render(
      <CollapsibleHeader
        title="Test Title"
        isOpen={false}
        clickable={false}
        onToggle={onToggleMock}
        testID="collapsible-header"
      />
    );
    const header = getByTestId('collapsible-header');
    fireEvent.press(header);
    expect(onToggleMock).not.toHaveBeenCalled();
  });

  it('renders with chevron down icon when open', () => {
    const { getByTestId } = render(
      <CollapsibleHeader title="Test Title" isOpen={true} onToggle={() => {}} testID="collapsible-header" />
    );
    expect(getByTestId('chevron-down-icon')).toBeTruthy();
  });

  it('renders with chevron right icon when closed', () => {
    const { getByTestId } = render(
      <CollapsibleHeader title="Test Title" isOpen={false} onToggle={() => {}} testID="collapsible-header" />
    );
    expect(getByTestId('chevron-right-icon')).toBeTruthy();
  });

  it('matches snapshot when open', () => {
    const { toJSON } = render(<CollapsibleHeader title="Test Title" isOpen={true} onToggle={() => {}} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when closed', () => {
    const { toJSON } = render(<CollapsibleHeader title="Test Title" isOpen={false} onToggle={() => {}} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
