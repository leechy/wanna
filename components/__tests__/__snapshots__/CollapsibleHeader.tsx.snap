// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleHeader matches snapshot when closed 1`] = `
<View
  style={
    {
      "borderRightColor": "transparent",
      "borderRightWidth": 1,
      "flex": 1,
      "margin": 0,
      "padding": 0,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 4,
        "justifyContent": "flex-start",
        "opacity": 1,
        "paddingBottom": 12,
        "paddingTop": 18,
      }
    }
  >
    <ChevronRightIcon
      color="#0000FF"
      height={24}
      testID="chevron-right-icon"
      width={24}
    />
    <Text
      style={
        [
          {
            "fontFamily": "Montserrat",
            "fontSize": 18,
            "fontWeight": "700",
            "lineHeight": 24,
          },
          {
            "color": "#0000FF",
          },
        ]
      }
    >
      Test Title
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "Montserrat",
            "fontSize": 16,
            "fontWeight": "500",
            "marginLeft": 8,
            "opacity": 0.5,
            "paddingTop": 2,
          },
          {
            "color": "#0000FF",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`CollapsibleHeader matches snapshot when open 1`] = `
<View
  style={
    {
      "borderRightColor": "transparent",
      "borderRightWidth": 1,
      "flex": 1,
      "margin": 0,
      "padding": 0,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 4,
        "justifyContent": "flex-start",
        "opacity": 1,
        "paddingBottom": 12,
        "paddingTop": 18,
      }
    }
  >
    <ChevronDownIcon
      color="#0000FF"
      height={24}
      testID="chevron-down-icon"
      width={24}
    />
    <Text
      style={
        [
          {
            "fontFamily": "Montserrat",
            "fontSize": 18,
            "fontWeight": "700",
            "lineHeight": 24,
          },
          {
            "color": "#0000FF",
          },
        ]
      }
    >
      Test Title
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "Montserrat",
            "fontSize": 16,
            "fontWeight": "500",
            "marginLeft": 8,
            "opacity": 0.5,
            "paddingTop": 2,
          },
          {
            "color": "#0000FF",
          },
        ]
      }
    />
  </View>
</View>
`;
